modules/programs/brew-bundle.nix

```diff
@@ -0,0 +1,225 @@
+        <option>programs.brew-bundle.brews</option>.
+
+        Note that you need to be signed into the Mac App Store for <command>mas</command> to
+        successfully install and upgrade applications, and that unfortunately apps removed from this
+        option will not be uninstalled automatically even if
+        <option>programs.brew-bundle.cleanup</option> is set to <literal>"uninstall"</literal>
+        or <literal>"zap"</literal> (this is currently a limitation of Homebrew Bundle).
+
+        For more information on <command>mas</command> see: https://github.com/mas-cli/mas
+      '';
+    };
+
+    whalebrews = mkOption {
+      type = with types; listOf str;
+      default = [];
+      example = [ "whalebrew/wget" ];
+      description = ''
+        Docker images to install using <command>whalebrew</command>.
+
+        When this option is used, <literal>"whalebrew"</literal> is automatically added to
+        <option>programs.brew-bundle.brews</option>.
+
+        For more information on <command>whalebrew</command> see:
+        https://github.com/whalebrew/whalebrew
+      '';
+    };
+
+    extraConfig = mkOption {
+      type = types.lines;
+      default = "";
+      example = ''
+        # 'brew tap' with custom Git URL
+        tap "user/tap-repo", "https://user@bitbucket.org/user/homebrew-tap-repo.git"
+
+        # set arguments for all 'brew cask install' commands
+        cask_args appdir: "~/Applications", require_sha: true
+
+        # 'brew install --with-rmtp', 'brew services restart' on version changes
+        brew "denji/nginx/nginx-full", args: ["with-rmtp"], restart_service: :changed
+        # 'brew install', always 'brew services restart', 'brew link', 'brew unlink mysql' (if it is installed)
+        brew "mysql@5.6", restart_service: true, link: true, conflicts_with: ["mysql"]
+
+        # 'brew cask install --appdir=~/my-apps/Applications'
+        cask "firefox", args: { appdir: "~/my-apps/Applications" }
+        # 'brew cask install' only if '/usr/libexec/java_home --failfast' fails
+        cask "java" unless system "/usr/libexec/java_home --failfast"
+      '';
+      description = "Extra lines to be added verbatim to the generated Brewfile.";
+    };
+  };
+
+  config = {
+    assertions = mkIf cfg.enable [
+      {
+        assertion = builtins.pathExists /usr/local/bin/brew;
+        message = ''
+          Homebrew not installed.
+
+          Please install Homebrew yourself before using the programs.brew-bundle module.
+
+          See installation instructions at: https://brew.sh
+        '';
+      }
+    ];
+
+    programs.brew-bundle.brews =
+      optional (cfg.masApps != {}) "mas" ++
+      optional (cfg.whalebrews != []) "whalebrew";
+
+    environment.variables = mkIf cfg.enable (
+      (if cfg.userConfig.brewfile then { HOMEBREW_BUNDLE_FILE = "${brewfile}"; } else {}) //
+      (if cfg.userConfig.noLock then { HOMEBREW_BUNDLE_NO_LOCK = "1"; } else {})
```

Comment:
Where does the lockfile go if this is disabled? And as a side remark it seems like this in general doesn't support a workflow with a lockfile which feels like the better approach to me.

Reply:
The lockfile is generated in the directory where the Brewfile is located whenever `brew bundle [install]` is run. So if this variable isn't set in the users environment, and the user runs `brew bundle [install]` themselves (with `hombrew.userConfig.brewfile` set to `true`, which is the current default), they'll see a warning like:

```
Warning: Could not write to /nix/store/gyg4yzd3ds5apzkcz211l6529ndfr0jc-Brewfile.lock.json!
```

Like `homebrew.userConfig.brewfile` this is another example of an option that I basically never expect to want to disable, but since I do sometimes run `brew bundle` myself (and want it to reference the generated Brewfile in the store by default), I do like that the module sets `HOMEBREW_BUNDLE_NO_LOCK` for me. I'd be happy to kill this option, and just add the variable to `environment.variables` when the module is enabled. I included it as an option out of an abundance of caution, since just like `homebrew.userConfig.brewfile`, I was worried that for some users, having that variable set in their environment would be annoying for some use-case of theirs.

After writing this comment and the one above, I am leaning towards killing both the `userConfig` options, and instead just adding both variables to `environment.variables` whenever the module is enabled.

@eraserhd, @kclejeune, what do you guys think? Would you be annoyed if this module just added those two variables to `environment.variables`, and didn't include an option to disable it doing so? Do you know of any common Homebrew Bundle use-cases where that would be annoying?

Reply:
I don't think that would complain on a single user installation and instead create an orphan lockfile in the store.

The main thing I was getting at is that I would expect the lockfile to be part of your configuration so it's actually used when switching the system. While the current approach seems to explicitly avoid it.

Reply:
To my knowledge, brew bundle doesn't actually make use of the lockfile when installing packages since there isn't really a notion of multiple package versions like we're used to in nixpkgs. 

I think you'd be safe to have this either be generated and kept in the nix store, or to just not be generated at all.

Reply:
> To my knowledge, brew bundle doesn't actually make use of the lockfile when installing packages since there isn't really a notion of multiple package versions like we're used to in nixpkgs.

Yeah, my understanding is that the lockfiles are generated for debugging purposes.

>  I think you'd be safe to have this either be generated and kept in the nix store, or to just not be generated at all.

Given that they're just for debugging, and that the lockfile would be generated during activation (which makes putting it in the store a hassle), I definitely don't think this module should generate them itself.

> Regarding where to set the variables associated with this, I don't really see an issue with putting them directly in the module as opposed to environment.variables since I can't think of much reason to change them in the first place, but don't have a strong preference either way.

To be clear, these don't need to be set at all in the activation script, since `brew bundle` is run with the `--no-lock` and `--file` flags in the script. The only effect of the `homebrew.userConfig` options was to govern the default behavior of the `brew bundle` command, if/when invoked by the user themselves.

---

Given the conversation here so far, I do still think that it's the right call to add these variables to `environment.variables`, since I think it's intuitive that enabling a module like this would result in the relevant command (in this case `brew bundle`) being configured to be aware and play well with the configuration this module generates. However, I'm now pretty convinced that having an option for these isn't the right call, since I've updated that people are very unlikely to want to change them from their current defaults, and options should only exist if they are going to be useful to people (i.e., some people will want to change them).

As such, I've removed the `homebrew.userConfig` options, and just setup the module to add these variables to `environment.variables` when this module is enabled.

Reply:
> > To my knowledge, brew bundle doesn't actually make use of the lockfile when installing packages since there isn't really a notion of multiple package versions like we're used to in nixpkgs.
> 
> 
> 
> Yeah, my understanding is that the lockfiles are generated for debugging purposes.
> 
> 
> 
> >  I think you'd be safe to have this either be generated and kept in the nix store, or to just not be generated at all.
> 
> 
> 
> Given that they're just for debugging, and that the lockfile would be generated during activation (which makes putting it in the store a hassle), I definitely don't think this module should generate them itself.
> 
> 
> 
> > Regarding where to set the variables associated with this, I don't really see an issue with putting them directly in the module as opposed to environment.variables since I can't think of much reason to change them in the first place, but don't have a strong preference either way.
> 
> 
> 
> To be clear, these don't need to be set at all in the activation script, since `brew bundle` is run with the `--no-lock` and `--file` flags in the script. The only effect of the `homebrew.userConfig` options was to govern the default behavior of the `brew bundle` command, if/when invoked by the user themselves.
> 
Oops, I totally overlooked this part. This makes sense. 
> 
> 
> ---
> 
> 
> 
> Given the conversation here so far, I do still think that it's the right call to add these variables to `environment.variables`, since I think it's intuitive that enabling a module like this would result in the relevant command (in this case `brew bundle`) being configured to be aware and play well with the configuration this module generates. However, I'm now pretty convinced that having an option for these isn't the right call, since I've updated that people are very unlikely to want to change them from their current defaults, and options should only exist if they are going to be useful to people (i.e., some people will want to change them).
> 
> 
> 
> As such, I've removed the `homebrew.userConfig` options, and just setup the module to add these variables to `environment.variables` when this module is enabled.

This sounds reasonable. Thanks for clarifying :)

Reply:
Sweet! Much appreciation for all you helpful comments/thoughts.